### PR TITLE
New Kubernetes Monitor

### PR DESCRIPTION
**Pull request by mheese**
Implements https://github.com/aporeto-inc/aporeto/issues/953 .

### CODE REVIEW

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "8184cd1b87a886a7dbf34d8cac739c22d42ded0e"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "db9dac4d3ca26a9f80ffb61d13ee9ad902bdb99a"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "8184cd1b87a886a7dbf34d8cac739c22d42ded0e"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "82e7936a0751edafc580c9942e778fe287414f48"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "4afd87fe2bfb773655ba5b4fa0d7106ced4591d5"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "8184cd1b87a886a7dbf34d8cac739c22d42ded0e"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "ce71d606b56a4e7c90a163926abbb03cd2f006d8"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "66571739e0aa8cb016fd5e71ed77b9e6bae93fcc"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "e8010239a4927a551318f35ebbbe8941bc8f7693"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "ce71d606b56a4e7c90a163926abbb03cd2f006d8"
  }
]
```

File: Gopkg.toml [hunk 1]
```diff
@@ -12,17 +12,45 @@ required = ["github.com/docker/distribution"]
...
+  version = "v0.1.10"
+
+#
+# Kubernetes dependencies
+# NOTE: always match exactly to what controller-runtime uses
+#
 [[constraint]]
   name = "k8s.io/api"
-  version = "kubernetes-1.10.0"
+  version = "kubernetes-1.13.1"
```
Comment (dstiliadis) [inline Gopkg.toml:29]: Just wondering if we have any backward compatibility issue with 1.11 if we stick to 1.13 clients.

File: monitor/internal/pod/controller.go [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+			if err := r.handler.Policy.HandlePUEvent(
+				handlePUCtx,
+				puID,
+				common.EventDestroy,
+				policy.NewPURuntimeWithDefaults(),
+			); err != nil {
+				zap.L().Error("failed to handle destroy event", zap.String("puID", puID), zap.Error(err))
+				return reconcile.Result{}, ErrHandlePUDestroyEventFailed
+			}
+			return reconcile.Result{}, nil
```
Comment (dstiliadis) [inline monitor/internal/pod/controller.go:152]: Can you add a note here, that we should just need one of the two events in engine and we need to fix that there.

**Review by dstiliadis — DISMISSED** [review]

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "77313ed8fcc1f1aeaf2f7f6e7fa4e73d95d9e398"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "b350e373cb5f9eb152fb0d5d2c58db27668fe1c8"
  }
]
```

File: Gopkg.toml (hunk 1 above)
Comment (mheese) [reply Gopkg.toml:29]: The operator is also using this specific version of kubernetes. For the controller-runtime it is very important to match the Kubernetes version to what the controller-runtime is using internally. I always match them against that for the least problems.

In the operator I tested everything down to 1.9, and have not seen any issues yet. We're not using any extravagant features in the controller in the monitor, so I don't think that there is going to be issues.

File: monitor/internal/pod/controller.go (hunk 2 above)
Comment (mheese) [reply monitor/internal/pod/controller.go:152]: will do

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "b350e373cb5f9eb152fb0d5d2c58db27668fe1c8"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "9bc6c9e7d7bfe326859e894de746236fe5a8b417"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "9bc6c9e7d7bfe326859e894de746236fe5a8b417"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "b350e373cb5f9eb152fb0d5d2c58db27668fe1c8"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "b350e373cb5f9eb152fb0d5d2c58db27668fe1c8"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "143815c54b5e3ac58e6aa869470ecc7b65ba8c93"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "143815c54b5e3ac58e6aa869470ecc7b65ba8c93"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "46cf24573c19b8451b51c6d8cbcf55170fb37dc8"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "73b2d5f780d0dabfb9d3ecfcf4f8e85f6026cdc0"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "020d450211c73ed7e62be8b9cd90200627d1625c"
  }
]
```

**Comment by mheese** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "new-kubernetes-monitor",
    "component": "trireme-lib",
    "pr-id": "750",
    "commit-sha": "73b2d5f780d0dabfb9d3ecfcf4f8e85f6026cdc0"
  },
  {
    "project": "new-kubernetes-monitor",
    "component": "enforcerd",
    "pr-id": "1049",
    "commit-sha": "bdac9c6ed669c676e34b2c2d7495d8f03d47a8b4"
  }
]
```

**Review by dstiliadis — APPROVED** [review]

**Pull request merged**